env/ipp_aes_ctr_provider.h

```diff
@@ -0,0 +1,89 @@
+//  Copyright (c) 2011-present, Facebook, Inc.  All rights reserved.
+//  Copyright (c) 2020 Intel Corporation
+//  This source code is licensed under both the GPLv2 (found in the
+//  COPYING file in the root directory) and Apache 2.0 License
+//  (found in the LICENSE.Apache file in the root directory).
+
+#pragma once
+
+#if !defined(ROCKSDB_LITE)
+
+#ifdef IPPCP
+
+#include <ippcp.h>
+
+#endif  // IPPCP
+
+#include <wmmintrin.h>
+
+#include <string>
+
+#include "env_encryption_ctr.h"
+
+namespace ROCKSDB_NAMESPACE {
+
+enum struct KeySize { AES_128 = 16, AES_192 = 24, AES_256 = 32 };
+
+#if defined(IPPCP) && (defined(HAVE_SSE42) || defined(HAVE_SSE2))
+
+class IppAESCTRCipherStream : public BlockAccessCipherStream {
+ public:
+  static const size_t kBlockSize = 16;    // in bytes
+  static const size_t kCounterLen = 128;  // in bits
+
+  IppAESCTRCipherStream(IppsAESSpec* aes_ctx, const char* init_vector);
+
+  virtual Status Encrypt(uint64_t fileOffset, char* data,
+                         size_t dataSize) override;
+  virtual Status Decrypt(uint64_t fileOffset, char* data,
+                         size_t dataSize) override;
+  virtual size_t BlockSize() override { return kBlockSize; }
+
+ protected:
+  virtual void AllocateScratch(std::string&) override {}
+  virtual Status EncryptBlock(uint64_t, char*, char*) override {
+    return Status::OK();
+  }
+  virtual Status DecryptBlock(uint64_t, char*, char*) override {
+    return Status::OK();
+  }
+
+ private:
+  IppsAESSpec* aes_ctx_;
+  __m128i init_vector_;
+};
+
+#endif  // IPPCP
+
+class IppAESCTRProvider : public CTREncryptionProvider {
+ public:
+  static Status CreateProvider(const std::string& id,
+                               IppAESCTRProvider** provider);
+
+#if defined(IPPCP) && (defined(HAVE_SSE42) || defined(HAVE_SSE2))
```

Comment:
Should this #ifdef be around everything or just the variables?  Seems like weird compilation rules otherwise...

Reply:
I wanted to avoid putting the #ifdef in EncryptionProvider::CreateFromString in env/env_encryption.cc. Will do that.

Reply:
I don't want the #ifdef in EncryptionProvider either.  But you still can define this class, just without the values that reference IPPCP headers.  Make them conditional but leave the class as valid perhaps?